lib/galaxy/model/__init__.py

```diff
@@ -644,6 +644,7 @@ def stderr(self, stderr):
 LEFT OUTER JOIN library_dataset_dataset_association ON dataset.id = library_dataset_dataset_association.dataset_id
 WHERE dataset.id IN (SELECT dataset_id FROM per_hist_hdas)
     AND library_dataset_dataset_association.id IS NULL
+    AND (dataset.object_store_id NOT LIKE 'user_objects://%' OR dataset.object_store_id IS NULL)
```

Comment:
the `user_objects://` prefix seems to be a source of truth (used to determine whether an object store is a user-defined source) that is specified in 2 places: here and `objectstore/__init__.py` (`_is_user_defined_source`). Can this be refactored so we have it in one place only?

Reply:
It is used in many more places, which is why I hesitated initially to refactor it as part of a bug fix, but I guess the refactoring is safe enough :sweat_smile: 

I pushed the refactoring in https://github.com/galaxyproject/galaxy/pull/19323/commits/f761a43c14977579a441fe04e63bd5a91a7aa742

Reply:
Thanks - definitely a big improvement!